akka-persistence-testkit/src/main/scala/akka/persistence/testkit/javadsl/UnpersistentBehavior.scala

```diff
@@ -0,0 +1,130 @@
+      behavior: Behavior[Command],
+      fromState: State,
+      fromOffset: Long): UnpersistentBehavior[Command, Event, State] = {
+    require(fromOffset >= 0, "fromOffset must be at least zero")
+
+    val fromStateAndOffset = Option(fromState).map(_ -> fromOffset)
+    val eventProbe = new PersistenceProbeImpl[Event]
+    val snapshotProbe = new PersistenceProbeImpl[State]
+
+    val b =
+      Unpersistent.eventSourced(behavior, fromStateAndOffset) { (event: Event, offset: Long, tags: ScalaSet[String]) =>
+          eventProbe.persist((event, offset, tags))
+        } { (snapshot, offset) =>
+          snapshotProbe.persist((snapshot, offset, ScalaSet.empty))
+        }
+
+    new UnpersistentBehavior(b, eventProbe.asJava, snapshotProbe.asJava)
+  }
+
+  def fromEventSourced[Command, Event, State](
+      behavior: Behavior[Command]): UnpersistentBehavior[Command, Event, State] =
+    fromEventSourced(behavior, null.asInstanceOf[State], 0)
+
+  def fromDurableState[Command, State](
+      behavior: Behavior[Command],
+      fromState: State): UnpersistentBehavior[Command, Void, State] = {
+    val probe = new PersistenceProbeImpl[State]
+    val b =
+      Unpersistent.durableState(behavior, Option(fromState)) { (state, version, tag) =>
+        probe.persist((state, version, if (tag == "") ScalaSet.empty else ScalaSet(tag)))
+      }
+
+    new UnpersistentBehavior(b, noEventProbe, probe.asJava)
+  }
+
+  def fromDurableState[Command, State](behavior: Behavior[Command]): UnpersistentBehavior[Command, Void, State] =
+    fromDurableState(behavior, null.asInstanceOf[State])
+
+  private val noEventProbe: PersistenceProbe[Void] =
+    new PersistenceProbe[Void] {
+      def getAllEffects(): List[PersistenceEffect[Void]] =
+        // could return an empty list, but the intent is that any use of this probe should fail the test
+        boom()
+
+      def extract(): PersistenceEffect[Void] = boom()
+      def expectPersistedClass[S <: Void](clazz: Class[S]): PersistenceEffect[S] = boom()
+      def hasEffects: Boolean = boom()
+      def expectPersisted(obj: Void): PersistenceProbe[Void] = boom()
+      def expectPersisted(obj: Void, tag: String): PersistenceProbe[Void] = boom()
+      def expectPersisted(obj: Void, tags: Set[String]): PersistenceProbe[Void] = boom()
+
+      private def boom() = throw new AssertionError("No events were persisted")
+    }
+}
+
+class UnpersistentBehavior[Command, Event, State] private (
+    behavior: Behavior[Command],
+    eventProbe: PersistenceProbe[Event],
+    stateProbe: PersistenceProbe[State]) {
+  def getBehavior(): Behavior[Command] = behavior
+
+  /** NB: durable state behaviors will not publish events to this probe */
+  def getEventProbe(): PersistenceProbe[Event] = eventProbe
+
+  def getStateProbe(): PersistenceProbe[State] = stateProbe
+  def getSnapshotProbe(): PersistenceProbe[State] = stateProbe
+}
+
+case class PersistenceEffect[T](persistedObject: T, offset: Long, tags: Set[String])
```

Comment:
Really not in love with this name... could make this common.

Reply:
For now, leaning toward unrelated and separate Java and Scala `PersistenceEffect`s and keeping it tuples on the inside.